### PR TITLE
CTPPS: DAQ mapping per interval-of-validity

### PR DESCRIPTION
**Pull request by jan-kaspar**
- `TotemDAQMappingESSourceXML` receives configuration per interval-of-validity.
- There are two mapping files:
  - `ctpps_mapping_to_fill_5288.xml`: before TS2
  - `ctpps_mapping_from_fill_5330.xml`: after TS2
- The association file - interval-of-validity is done in `EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py`
- A config file to test the configuration switching: `EventFilter/CTPPSRawToDigi/test/global_daq_test.py`

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
I guess `CondFormats/CTPPSReadoutObjects`needs be assigned to `alca` and `db` @slava77 @smuzaffar

**Comment by slava77** [comment]
On 10/3/16 6:20 AM, Marco Musich wrote:
> I guess |CondFormats/CTPPSReadoutObjects|needs be assigned to |alca| and
> |db| @slava77 <https://github.com/slava77> @smuzaffar
> <https://github.com/smuzaffar>
> 

cms-bot/categories.py can be edited to achieve this
https://github.com/cms-sw/cms-bot/blob/master/categories.py
edit it online and submit a PR with a change

**Comment by jan-kaspar** [comment]
I've gone through some of the error reports and each time the problem comes from an attempt to get ES data in event `run: 1 lumi: 0 event: 0`. I've tried to adjust the python config to cover also for this event id, but I've failed. I've tried the following
  * `"1:0:0"`: assertion failed, LB must be grater than 0
  * `"1:min"`: this gives `run: 1 lumi: 0 event: 1`
  * `"1:0"`: this gives `run: 1 lumi: 0 event: 1`, again.
Is there another way how to pass "1:0:0" via the config?

Another thought: "1:0:0" looks rather artificial for a job processing real data, so I assume this is rather for a simulation. And for simulation our raw-to-digi makes no sense. In contrary, real data processing, the event id should be well within the range selectable in a python config. Am I correct?

**Comment by slava77** [comment]
On 10/3/16 9:29 AM, jan-kaspar wrote:
> I've gone through some of the error reports and each time the problem
> comes from an attempt to get ES data in event |run: 1 lumi: 0 event: 0|.
> I've tried to adjust the python config to cover also for this event id,
> but I've failed. I've tried the following
> 
>   * |"1:0:0"|: assertion failed, LB must be grater than 0
>   * |"1:min"|: this gives |run: 1 lumi: 0 event: 1|
>   * |"1:0"|: this gives |run: 1 lumi: 0 event: 1|, again. Is there
>     another way how to pass "1:0:0" via the config?
> 
> Another thought: "1:0:0" looks rather artificial for a job processing
> real data, so I assume this is rather for a simulation. And for
> simulation our raw-to-digi makes no sense. In contrary, real data
> processing, the event id should be well within the range selectable in a
> python config. Am I correct?

yes.

I suppose you can use 1:1:1 as a special case and code some logic to use
beginOfTime for it

**Comment by jan-kaspar** [comment]
> I suppose you can use 1:1:1 as a special case and code some logic to use
> beginOfTime for it

I can do this. I just thought that there was a cleaner and more standard way of doing it.

**Comment by jan-kaspar** [comment]
>  Another thought: "1:0:0" looks rather artificial for a job processing
>  real data, so I assume this is rather for a simulation. And for
>  simulation our raw-to-digi makes no sense. In contrary, real data
>  processing, the event id should be well within the range selectable in a
>  python config. Am I correct?

What I wanted to say by this was that an alternative solution would be simply not to run the CTPPS raw-to-digi for simulation and other jobs where it makes no sense. But I don't know how easy/difficult that would be.

**Comment by slava77** [comment]
On 10/3/16 11:27 AM, jan-kaspar wrote:
>     Another thought: "1:0:0" looks rather artificial for a job processing
>     real data, so I assume this is rather for a simulation. And for
>     simulation our raw-to-digi makes no sense. In contrary, real data
>     processing, the event id should be well within the range selectable in a
>     python config. Am I correct?
> 
> What I wanted to say by this was that an alternative solution would be
> simply not to run the CTPPS raw-to-digi for simulation and other jobs
> where it makes no sense. But I don't know how easy/difficult that would be.

The plan was to have simulation.
I'd rather not make data vs MC setup changes, if it is rather temporary.

**Comment by jan-kaspar** [comment]
> The plan was to have simulation.
> I'd rather not make data vs MC setup changes, if it is rather temporary.

But do you simulate RAW data? In TOTEM we have always simulated only DIGI.

**Comment by slava77** [comment]
On 10/3/16 11:33 AM, jan-kaspar wrote:
>     The plan was to have simulation.
>     I'd rather not make data vs MC setup changes, if it is rather temporary.
> 
> But do you simulate RAW data? In TOTEM we have always simulated only DIGI.

DIGI->RAW step will need to be implemented as a part of simulation

**Comment by jan-kaspar** [comment]
> I suppose you can use 1:1:1 as a special case and code some logic to use
> beginOfTime for it

Done. I've run some of the previously failing matrix and add-on tests and they have all passed now.

**Comment by slava77** [comment]
@cmsbuild please test

File: EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
 from CondFormats.CTPPSReadoutObjects.TotemDAQMappingESSourceXML_cfi import *
-TotemDAQMappingESSourceXML.mappingFileNames.append("CondFormats/CTPPSReadoutObjects/xml/ctpps_210_mapping.xml")
+totemDAQMappingESSourceXML.configuration = cms.untracked.VPSet(
+  cms.PSet(
+    validityRange = cms.untracked.EventRange("1:min - 280385:max"),
+    mappingFileNames = cms.untracked.vstring("CondFormats/CTPPSReadoutObjects/xml/ctpps_mapping_to_fill_5288.xml"),
+    maskFileNames = cms.untracked.vstring()
+  ),
+  cms.PSet(
+    validityRange = cms.untracked.EventRange("281601:min - 999999999:max"),
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py]: what will happen for runs 280386 to 281600?
I'd expect that the coverage should be continuous.

File: CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc [hunk 2]
```diff
@@ -208,44 +222,65 @@ const string TotemDAQMappingESSourceXML::tagT1ChannelType="t1_channel_type";
...
   verbosity(conf.getUntrackedParameter<unsigned int>("verbosity", 0)),
-  mappingFileNames(conf.getUntrackedParameter< vector<string> >("mappingFileNames")),
-  maskFileNames(conf.getUntrackedParameter< vector<string> >("maskFileNames"))
+  currentBlock(0),
+  currentBlockValid(false)
 {
+  for (const auto it : conf.getUntrackedParameter<vector<ParameterSet>>("configuration"))
+  {
+    ConfigBlock b;
+    b.validityRange = it.getUntrackedParameter<EventRange>("validityRange");
```
Comment (slava77) [inline CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc]: configuration which affects physics results of the algorithm should be tracked

**Comment by jan-kaspar** [comment]
@slava77

Relevant parameters changed to tracked.

The runs 280386 to 281600 correspond to TS2 where no physics data were taken there. Nevertheless, I've defined this time interval, too, for the `TotemDAQMappingESSourceXML` not to throw exceptions.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ggovi** [comment]
+1

File: EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py [hunk 3]
```diff
@@ -1,7 +1,26 @@
...
+  # before TS2 (2016)
+  cms.PSet(
+    validityRange = cms.EventRange("1:min - 280385:max"),
+    mappingFileNames = cms.vstring("CondFormats/CTPPSReadoutObjects/xml/ctpps_mapping_to_fill_5288.xml"),
+    maskFileNames = cms.vstring()
+  ),
+  # during TS2 (2016)
+  cms.PSet(
+    validityRange = cms.EventRange("280386:min - 281600:max"),
+    mappingFileNames = cms.vstring(),
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py]: what will be a result of having no mapping file?
The job should not crash.

Comment (slava77) [inline EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py]: we have some data in Run2016H-v1 in this range of runs in physics PDs.
If some reprocessing ends up triggered for it (privately or centrally), it should be able to run.

Comment (jan-kaspar) [inline EventFilter/CTPPSRawToDigi/python/totemRawToDigi_cff.py]: In `mappingFileNames` is empty, the corresponding map in EventSetup will be empty. Then, if the CTPPS raw-to-digi is executed and if there are some data from CTPPS FEDs (which should not be the case), these data are ignored:
https://github.com/CTPPS/cmssw/blob/ctpps_conf_daq_mapping/EventFilter/CTPPSRawToDigi/src/RawToDigiConverter.cc#L73
No crash.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16074 f4949cc
- code change is as described: to use different xml text files in the RP DAQ mapping ES producer
- jenkins tests pass and comparisons with baseline show no differences
- tests with recent data were done in 80X (see #16115)

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16074 efd417a
- after DAQ mapping file update

**Comment by ggovi** [comment]
+1

File: CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc [hunk 4]
```diff
@@ -291,6 +336,22 @@ edm::ESProducts< boost::shared_ptr<TotemDAQMapping>, boost::shared_ptr<TotemAnal
 
 //----------------------------------------------------------------------------------------------------
 
+DOMDocument* TotemDAQMappingESSourceXML::GetDOMDocument(string file)
+{
+  XercesDOMParser* parser = new XercesDOMParser();
```
Comment (davidlange6) [inline CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc]: @jan-kaspar - it looks like this misses XMLPlatformUtils::Initialize(); and terminate calls and leaks the parser. is all of that done somewhere else?

Comment (jan-kaspar) [inline CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc]: The Initialize and Terminate calls are here:
https://github.com/CTPPS/cmssw/blob/ctpps_conf_daq_mapping/CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc#L313
https://github.com/CTPPS/cmssw/blob/ctpps_conf_daq_mapping/CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc#L331

The "parser" really doesn't seem to be deleted. I will fix it.

Comment (jan-kaspar) [inline CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc]: OK, now parser is also deleted:
https://github.com/CTPPS/cmssw/blob/ctpps_conf_daq_mapping/CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc#L339
If you like this solution, I will backport it also to 8_0_X. Please let me know. Thanks!

**Comment by mmusich** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16074 a8e82da
- after update to non-leaking version of TotemDAQMappingESSourceXML::ParseXML
    - previously signed in https://github.com/cms-sw/cmssw/pull/16074#issuecomment-253859953
- jenkins tests pass and comparisons show no differences as expected (2016B workflow 136.731 has CTPPS data)

**Comment by mmusich** [comment]
+1

**Comment by slava77** [comment]
@ggovi please check this PR, it needs your signature
Thank you.

File: CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc [hunk 5]
```diff
@@ -82,11 +82,25 @@ class TotemDAQMappingESSourceXML: public edm::ESProducer, public edm::EventSetup
 private:
```
Comment (jan-kaspar) [inline CondFormats/CTPPSReadoutObjects/plugins/TotemDAQMappingESSourceXML.cc]: Sorry, I can't comment, I have no clue who is supposed to sign what.

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**